packages/vats/src/orchestration.js

```diff
@@ -0,0 +1,268 @@
+/**
+ * @template {keyof OrchestrationPowers} K
+ * @param {PowerStore} powers
+ * @param {K} name
+ */
+const getPower = (powers, name) => {
+  powers.has(name) || Fail`need powers.${bare(name)} for this method`;
+  return /** @type {OrchestrationPowers[K]} */ (powers.get(name));
+};
+
+export const ChainAccountI = M.interface('ChainAccount', {
+  getAccountAddress: M.call().returns(M.string()),
+  getLocalAddress: M.call().returns(M.string()),
+  getRemoteAddress: M.call().returns(M.string()),
+  getPort: M.call().returns(M.remotable('Port')),
+  executeTx: M.callWhen(M.arrayOf(M.record())).returns(M.any()),
+  executeEncodedTx: M.callWhen(M.string()).returns(M.any()),
+  deposit: M.callWhen(M.remotable('Payment'))
+    .optional(M.pattern())
+    .returns(AmountShape),
+  getPurse: M.callWhen(BrandShape).returns(M.remotable('VirtualPurse')),
+  close: M.callWhen().returns(M.promise()),
+});
+
+export const ConnectionHandlerI = M.interface('ConnectionHandler', {
+  onOpen: M.callWhen(M.any(), M.string(), M.string(), M.any()).returns(M.any()),
+  onClose: M.callWhen(M.any(), M.string()).returns(M.any()),
+  onReceive: M.callWhen(M.any(), M.string()).returns(M.any()),
+});
+
+/** @param {import('@agoric/base-zone').Zone} zone */
+const prepareChainAccount = zone =>
+  zone.exoClassKit(
+    'ChainAccount',
+    { account: ChainAccountI, connectionHandler: ConnectionHandlerI },
+    /**
+     * @param {Port} port
+     * @param {string} requestedRemoteAddress
+     */
+    (port, requestedRemoteAddress) =>
+      /**
+       * @type {{
+       *   port: Port;
+       *   connection: Connection | undefined;
+       *   localAddress: string | undefined;
+       *   requestedRemoteAddress: string;
+       *   remoteAddress: string | undefined;
+       *   accountAddress: ChainAddress | undefined;
+       * }}
+       */ (
+        harden({
+          port,
+          connection: undefined,
+          requestedRemoteAddress,
+          remoteAddress: undefined,
+          accountAddress: undefined,
+          localAddress: undefined,
+        })
+      ),
+    {
+      account: {
+        getAccountAddress() {
+          const { accountAddress } = this.state;
+          accountAddress || Fail`account address not available`;
+          return accountAddress;
+        },
+        getLocalAddress() {
+          const { localAddress } = this.state;
+          localAddress || Fail`local address not available`;
+          return localAddress;
+        },
+        getRemoteAddress() {
+          const { remoteAddress } = this.state;
+          remoteAddress || Fail`remote address not available`;
+          return remoteAddress;
+        },
+        getPort() {
+          return this.state.port;
+        },
+        /** @param {Bytes} packetBytes */
+        async executeEncodedTx(packetBytes) {
+          const { connection } = this.state;
+          connection || Fail`connection not available`;
+          // @ts-expect-error Fail does not satisfy undefined
```

Comment:
TS doesn't know that `Fail` throws
```suggestion
          if (!connection) throw Fail`connection not available`;
```